Helper function to plot residuals in partial correlation scatter plots

coopa33 opened the pull request:
Solution for issue #3507:

The scatter plots for the correlation analysis did not show residuals of X and Y regressed on partial-out variables Z for partial correlation. Which it should.

This was not only the case for the datapoints, but also for the marginal distributions, which also did not change when one or more partial-out variables were added. 
However, the plotted statistics did change while the plots did not.

**Original Analysis - No partial correlation**
<img width="1645" height="1405" alt="Screenshot From 2025-08-04 12-41-24" src="https://github.com/user-attachments/assets/9ca6d870-664a-43c5-97b7-bdc21636f6d1" />

**Original Analysis - Partial correlation**
<img width="1645" height="1405" alt="Screenshot From 2025-08-04 12-48-25" src="https://github.com/user-attachments/assets/8bc88b94-4bab-429a-8831-f3686f1a9cc8" />

### Problem

- The problem seemed to be that the functions .corrPairwisePlot() and .corrMatrixPlot() in "R/correlation.R" do not account for the presence of variables to be partialled out. 
- In there, the original data is fed to further plotting functions .corrScatter() and .corrMarginalDistribution(), which is why the scatterplot and marginal distributions don't change. 
- However, the function to plot the statistics .corrValuePlot() outputs the correct statistic, because it accesses the pre-computed corrResults where the partial-out variable is accounted for in calculating the statistic. 

### Solution

- To solve this, I defined a helper function called .corrPartialResiduals(). Depending on if options$partialOutVariables has any entries, it will either return the original data (as before), or it will return residual data X and Y regressed on one or more variables Z to be partialled out. 
- This allows me to only make changes in the higher-level functions .corrPairwisePlot() and .corrMatrixPlot(). Any subsequent functions will then use either the original data or the residuals and do not require further change

**Modified Analysis - No partial correlation**
<img width="1645" height="1405" alt="Screenshot From 2025-08-04 13-01-04" src="https://github.com/user-attachments/assets/2f0d4dd2-f77a-4676-bd18-f649226de96d" />

**Modified Analysis - Partial correlation**
<img width="1645" height="1405" alt="Screenshot From 2025-08-04 13-01-42" src="https://github.com/user-attachments/assets/6fb653b8-fab6-4321-a1ef-496f3277b646" />

**Modified Analysis - Multiple variables to partial out + Confidence & Prediction Intervals**
<img width="2502" height="1412" alt="Screenshot From 2025-08-04 13-03-50" src="https://github.com/user-attachments/assets/8ac70f1f-7105-433f-bee0-8b409b45b247" />

### Further Notes
- Please note that the residuals displayed are not rank-transformed as you would expect for Spearman's rho. However, this was also not a feature before either. But as the interface of .corrPartialResiduals includes the options parameter, this could be implemented if the Spearman option is chosen.  Then the function could return rank-transformed residuals.

coopa33 commented:
@JohnnyDoorn Hi Johnny, I'm not able to request a review from you here, probably because I'm not a collaborator for this module. Would you be able to review my fix for the issue?

mupeker commented:
When you place the two graphs side by side, you can provide the following explanation (perhaps in the help section):

"The normal correlation graph shows the relationship between the raw (original) values of the X and Y variables. The partial correlation graph reflects the relationship between X and Y after the effect of the Z variable has been removed. Therefore, the axes of the partial correlation graph represent the “residual” values obtained from regression models rather than the original variable values. As seen, when the effect of the Z variable is controlled, the relationship between X and Y exhibits a [strengthened/weakened/changed direction] structure."

coopa33 commented:
I have added a caption under the scatterplot, stating explicitly that the axes represent residuals:
<img width="845" height="625" alt="Screenshot From 2025-08-08 01-25-04" src="https://github.com/user-attachments/assets/f00b9aaf-0731-4c56-b5fe-7bfb4d4d3ce2" />

This also works with multiple variables to be controlled for:
<img width="847" height="767" alt="Screenshot From 2025-08-08 01-26-16" src="https://github.com/user-attachments/assets/49d6c8ed-9377-432d-beae-16f3ef0f7087" />

However, the function I use to do this unfortunately doesn't work with jaspGraphs::ggMatrixPlot objects (yet). 
So, for matrix plots, it displays the caption under every scatterplot instead of only once under the complete matrix. This can  look quite cluttered, but can be fixed by adjusting the size of the plot (this seems to work because within the matrix, the scatterplots are individual jaspPlotR objects):
<img width="1707" height="1408" alt="Screenshot From 2025-08-08 01-30-34" src="https://github.com/user-attachments/assets/ecbedeca-9f0e-4d99-8d96-a044c973a112" />

But for the matrix format (if display pairwise is not clicked), it cannot be adjusted and just looks cluttered, even when making the matrix object bigger. Also, if the variables controlled for are identical for every combination, it is unnecessary to caption every plot:
<img width="2157" height="1395" alt="Screenshot From 2025-08-08 01-40-29" src="https://github.com/user-attachments/assets/f0872e03-fee5-41e8-8e44-3c7b99f6d839" />

I'll work on making the function work with ggMatrixPlot objects. If anyone has some input on how to make this work, pls let me know :)

@mupeker 
- I've added additional description in the help section. The same should also now be visible in the information box that appears if you hover over the plot option. 
- Is the caption explicit enough and a good idea? Otherwise, I was also thinking we could also have the X/Y labels state that with some prefix of sorts (like Residual: "variableName") but that might be even more cluttered then?

mupeker commented on inst/help/Correlation.md:
```diff
@@ -46,7 +46,7 @@ The Correlation analysis allows estimation of the population correlation, as wel
 
 #### Plots
-- Scatter plots: Display a scatter plots for each possible combination of the selected variables. In a matrix format, these are placed above the diagonal.
+- Scatter plots: Displays scatter plots for all variable pairs. In a matrix format, these are placed above the diagonal. For partial correlations, plots show the relationship between X and Y after removing the effect of Z, and axes then represent residuals from regressing X and Y on Z.
```
It is suggested to replace the conjunction "and" with "Therefore" in the help text of JASP, specifically in the sentence:

 "For partial correlations, plots show the relationship between X and Y after removing the effect of Z**, and** axes then represent residuals from regressing X and Y on Z."
 
"For partial correlations, plots show the relationship between X and Y after removing the effect of Z### . Therefore, axes then represent residuals from regressing X and Y on Z."
This modification improves the flow of the sentence and clearly highlights that the second part of the statement is a logical consequence of the first. This makes the help text more explanatory and easier for users to understand the core concept behind partial correlation plots.

coopa33 replied:
Good catch, in that case we could make it even more explicit? As in:  "In that case, axes represent residuals from regressing X and Y on Z, instead of the raw X and Y variables", or something of the sort?

mupeker replied:
Thank you for the quick response. I agree with your suggestion to make the wording more explicit. Your proposed phrase clarifies the difference between the residuals and the raw variables, which is the key point for a correct interpretation of the plot. Your suggested wording would be a significant improvement.

JohnnyDoorn commented:
Really great stuff @coopa33! One suggestion for clarifying that it concerns the residuals is to tweak the name of the plot (the title that is listed right below "Scatter plots":
"petal.length vs. sepal.length" -> "Residuals of petal.length vs. sepal.length, after controlling for petal.width", although that could become quite lengthy and I think it's already good as is with the residuals on the axis labels (especially if we improve it after the patchwerk update).

coopa33 commented:
@JohnnyDoorn Then if we tweak the title to explicitly state that these are residuals, we could also shorten the axes labels and make it less cluttered (Res: <varName>). I will try out if it works and how it looks and let you know. Thanks!

JohnnyDoorn commented:
Great work @coopa33!
I took a look at how it looks for linear regression (partial plots), and there we use "Residuals <var name>" on the axes, so maybe you can also do that here, for consistency? I do like that "controlling for.." is included in correlation, so I will look into also including that in the plot names for linear regression.

coopa33 commented:
@JohnnyDoorn Glad to help!
I'll take a look!

JohnnyDoorn commented:
Yes, awesome!

coopa33 commented:
@JohnnyDoorn also a quick question, I'm noting that in lin regression a partial plot is also produced for only one covariate/predictor. I'm just thinking because then it should be just a scatterplot of the raw variables and not residuals right? If that is the case. should we control flow it not show the "Residuals" prefix? Or maybe the plot should not display for only one predictor?

coopa33 commented on R/regressionlinear.R:
```diff
@@ -746,7 +746,12 @@ RegressionLinearInternal <- function(jaspResults, dataset = NULL, options) {
   }
 
-  if (options$dependent != "" && length(predictors) > 0) {
+  if(options$dependent != "" && length(predictors) == 1) {
```
Temporary solution for single predictor partial plot. In this case, error message displays that partial plots require at least two predictors
